### PR TITLE
ESM2 Tutorial Updates

### PR DESCRIPTION
**Pull request by farhadrgh**
- Deprecate `infer.py` example for finetuning and add Inference tutorial. 
- Add [ESM2 Mutant Design](https://github.com/NVIDIA/bionemo-framework/blob/bionemo1/docs/bionemo/notebooks/esm2nv-mutant-design.ipynb) notebook.

### CODE REVIEW

**Comment by jstjohn** [comment]
@farhadrgh would it be possible to follow the general patterns of https://github.com/NVIDIA/bionemo-framework/blob/main/docs/docs/user-guide/examples/bionemo-geneformer/geneformer-celltype-classification.ipynb and turn this into a notebook?

**Comment by farhadrgh** [comment]
[16080cd](https://github.com/NVIDIA/bionemo-framework/pull/426/commits/16080cdbb9088f1c2d96e966b7438d5a163829ab) allows partial batch inference initially suggested in https://github.com/NVIDIA/bionemo-framework/pull/304 which was pending NeMo/Megatron version bump

File: docs/docs/user-guide/examples/bionemo-esm2/finetune.md [hunk 1]
```diff
@@ -174,31 +174,33 @@ data = [(seq, len(seq)/100.0) for seq in artificial_sequence_data]
...
 ```bash
-python /workspace/bionemo2/sub-packages/bionemo-esm2/src/bionemo/esm2/model/finetune/train.py
+python -m bionemo.esm2.model.finetune.train
 ```
+
 ## Notes
-1. The above example is fine-tuning a randomly initialized ESM-2 model for demonstration purposes. In order to fine-tune a pre-trained ESM-2 model, please download the ESM-2 650M checkpoint from NGC resources using the following bash command
+1. The above example is fine-tuning a 650M ESM-2 model. The pre-trained checkpoints can be downloaded from NGC resources using either the following bash command or the `load` function in `bionemo.core.data.load` as shown above.
     ```bash
     download_bionemo_data esm2/650m:2.0 --source ngc
```
Comment (jstjohn) [inline docs/docs/user-guide/examples/bionemo-esm2/finetune.md:205]: you no longer need to tell people to add the `--source ngc` part if you do not want to, it's default now.

File: sub-packages/bionemo-esm2/src/bionemo/esm2/model/finetune/train.py [hunk 2]
```diff
@@ -166,24 +170,20 @@ def train_model(
...
+    config = ESM2FineTuneSeqConfig(initial_ckpt_path=str(pretrain_ckpt_path))
+
+    checkpoint, metrics, trainer = train_model(
+        experiment_name=experiment_name,
+        experiment_dir=Path(experiment_results_dir),  # new checkpoint will land in a subdir of this
+        config=config,  # same config as before since we are just continuing training
+        data_module=data_module,
+        n_steps_train=n_steps_train,
+    )
+    print(f"Experiment completed with checkpoint stored at {checkpoint}")
```
Comment (jstjohn) [inline sub-packages/bionemo-esm2/src/bionemo/esm2/model/finetune/train.py:189]: One idea, if you print these two messages as two prints, then the last output line of the script will be the checkpoint. You can use that to further automate your notebooks/examples. For example:
```
print("Experiment completed with checkpoint stored at ")
print(f"{checkpoint}")
```
Then in your notebook you can have something like:
```
train_out:list[str] = !python -m bionemo.esm2.model.finetune.train [args] [whatever]
ckpt_path = train_out[-1]
```

Comment (farhadrgh) [reply sub-packages/bionemo-esm2/src/bionemo/esm2/model/finetune/train.py:189]: I am going to try using the pretrain script to run finetuning and deprecate this script in another PR

**Review by jstjohn — APPROVED** [review]

File: sub-packages/bionemo-esm2/src/bionemo/esm2/model/finetune/train.py (hunk 2 above)
Comment (jstjohn) [reply sub-packages/bionemo-esm2/src/bionemo/esm2/model/finetune/train.py:189]: great, either way printing the checkpoint as the last step would get you this ability.

Comment (jstjohn) [reply sub-packages/bionemo-esm2/src/bionemo/esm2/model/finetune/train.py:189]: so maybe we should do that in `train.py` in general?

File: docs/docs/user-guide/examples/bionemo-esm2/inference.ipynb [hunk 3]
```diff
@@ -0,0 +1,500 @@
+{
+ "cells": [
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "[![ Click here to deploy.](https://uohmivykqgnnbiouffke.supabase.co/storage/v1/object/public/landingpage/brevdeploynavy.svg)](https://console.brev.dev/launchable/deploy?launchableID=env-2p8Dz9VJ2kFSrTQMXVRtUMgymKx)\n",
+    "\n",
+    "NOTE: it takes about 10 minutes to deploy this notebook as a Launchable. As of this writing, we are working on a free tier so a credit card may be required. You can reach out to your NVIDIA rep for credits."
```
Comment (jstjohn) [inline docs/docs/user-guide/examples/bionemo-esm2/inference.ipynb:9]: can you change this to the nice boxed note? I'm going to do that as well.

File: sub-packages/bionemo-esm2/src/bionemo/esm2/model/finetune/train.py (hunk 2 above)
Comment (farhadrgh) [reply sub-packages/bionemo-esm2/src/bionemo/esm2/model/finetune/train.py:189]: yes, we definitely need that for finetuning, I will add that to pre-train script

File: docs/docs/user-guide/examples/bionemo-esm2/finetune.md [hunk 4]
```diff
@@ -219,21 +221,43 @@ python /workspace/bionemo2/sub-packages/bionemo-esm2/src/bionemo/esm2/model/fine
 3. We are using a small dataset of artificial sequences as our fine-tuning data in this example. You may experience over-fitting and observe no change in the validation metrics.
 
 # Fine-Tuned ESM-2 Model Inference
-Once we have a checkpoint we can create a config object by pointing the path in `initial_ckpt_path` and use that for inference. Since we need to load all the parameters from this checkpoint (and don't skip the head) we reset the `nitial_ckpt_skip_keys_with_these_prefixes` in this config. Now we can use the ```bionemo.esm2.model.fnetune.train.infer``` to run inference on prediction dataset.
+Now we can use the ```bionemo.esm2.model.fnetune.train.infer``` to run inference on an example prediction dataset.
```
Comment (pstjohn) [inline docs/docs/user-guide/examples/bionemo-esm2/finetune.md:224]:
```suggestion
Now we can use ```bionemo.esm2.model.finetune.train.infer``` to run inference on an example prediction dataset.
```

File: docs/docs/user-guide/examples/bionemo-esm2/finetune.md [hunk 5]
```diff
@@ -219,21 +221,43 @@ python /workspace/bionemo2/sub-packages/bionemo-esm2/src/bionemo/esm2/model/fine
...
+This will create a result `.pt` file under `$WORKDIR/esm2_finetune_tutorial/inference_results.pt` which can be loaded via PyTorch library in python environment:
 
-```bash
-python /workspace/bionemo2/sub-packages/bionemo-esm2/src/bionemo/esm2/model/finetune/infer.py
+```python
+import torch
+
+# Set the path to results file e.g. /workspace/bionemo2/esm2_finetune_tutorial/inference_results.pt
+# results_path = /workspace/bionemo2/esm2_finetune_tutorial/inference_results.pt
+results = torch.load(results_path)
```
Comment (pstjohn) [inline docs/docs/user-guide/examples/bionemo-esm2/finetune.md:249]: Is this just to avoid running inference in a notebook? if we're trying to be a library, having to invoke this as a script and then dumping the results to a file seems like an odd API

File: docs/docs/user-guide/examples/bionemo-esm2/mutant-design.ipynb
Comment (pstjohn) [inline docs/docs/user-guide/examples/bionemo-esm2/mutant-design.ipynb:1]: Can you somehow highlight the input sequence in the Positional Token Probabilities plot?

File: sub-packages/bionemo-core/src/bionemo/core/data/resources/esm2.yaml [hunk 6]
```diff
@@ -40,3 +40,11 @@
   sha256: 006911f92bbc0ded7ea302bbdbfab4c694b409e699c32fd49de1c527a99dba3e  # pragma: allowlist secret
   owner: Peter St John <pstjohn@nvidia.com>
   description: Test data for ESM2 pretraining.
+
+- tag: esm2_inference_testdata:2.0
+  ngc: nvidia/clara/esm2_inference_testdata:2.0  # TODO: upload to NGC
```
Comment (pstjohn) [inline sub-packages/bionemo-core/src/bionemo/core/data/resources/esm2.yaml:45]: is this done? will CI pass without it being uploaded?

File: sub-packages/bionemo-llm/src/bionemo/llm/model/biobert/model.py [hunk 7]
```diff
@@ -453,6 +453,8 @@ def forward(
         output = {"token_logits": logits, "binary_logits": binary_logits}
         if self.include_hiddens:
             output["hidden_states"] = hidden_states.transpose(0, 1).contiguous()  # [s b h] => [b s h]
+        if self.include_hiddens or not self.skip_logits:
+            output["input_ids"] = input_ids
```
Comment (pstjohn) [inline sub-packages/bionemo-llm/src/bionemo/llm/model/biobert/model.py:457]: why do we want this? just passing back out the inputs?

Does this just make it easier to dump the inference results to a file and include the inputs? If so, I think it's worth coming up with another way to do that, I'm worried this might mess up some other training code or otherwise increase GPU memory allocation by passing around tensors we no longer need

File: sub-packages/bionemo-core/src/bionemo/core/data/resources/esm2.yaml (hunk 6 above)
Comment (farhadrgh) [reply sub-packages/bionemo-core/src/bionemo/core/data/resources/esm2.yaml:45]: yes, this was already uploaded

File: docs/docs/user-guide/examples/bionemo-esm2/mutant-design.ipynb
Comment (farhadrgh) [reply docs/docs/user-guide/examples/bionemo-esm2/mutant-design.ipynb:1]: I think this is the plot they included in the paper and reference bionemo. This was made by AAlpha and I tried to port it over just to make it compatible with the v2.0

File: docs/docs/user-guide/examples/bionemo-esm2/finetune.md (hunk 5 above)
Comment (farhadrgh) [reply docs/docs/user-guide/examples/bionemo-esm2/finetune.md:249]: yeah, I see your point, ideally we would refactor and make these inference-related scripts a CLI wrapper around the forward call. I will add that to TODO for future release

File: sub-packages/bionemo-llm/src/bionemo/llm/model/biobert/model.py (hunk 7 above)
Comment (farhadrgh) [reply sub-packages/bionemo-llm/src/bionemo/llm/model/biobert/model.py]: I just added an argument to `include_input_ids` optionally per our discussion

**Review by pstjohn — APPROVED** [review]
LGTM, thanks!
